connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -117,10 +129,91 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
     responseObserver.onCompleted()
   }
 
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val spark = dataframe.sparkSession
+    val schema = dataframe.schema
+    // TODO: control the batch size instead of max records
+    val maxRecordsPerBatch = spark.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = spark.sessionState.conf.sessionLocalTimeZone
+
+    SQLExecution.withNewExecutionId(dataframe.queryExecution, Some("collectArrow")) {
+      val pool = ThreadUtils.newDaemonSingleThreadExecutor("connect-collect-arrow")
+      val tasks = collection.mutable.ArrayBuffer.empty[Future[_]]
+      val rows = dataframe.queryExecution.executedPlan.execute()
+
+      if (rows.getNumPartitions > 0) {
+        val batches = rows.mapPartitionsInternal { iter =>
+          ArrowConverters
+            .toArrowBatchIterator(iter, schema, maxRecordsPerBatch, timeZoneId)
+        }
+
+        val processPartition = (iter: Iterator[(Array[Byte], Long, Long)]) => iter.toArray
+
+        val resultHandler = (partitionId: Int, taskResult: Array[(Array[Byte], Long, Long)]) => {
```

Comment:
I think we should discuss client side ordering of results a bit more. My main problem with this is is that it is unexpected behavior, undocumented behavior (hint hint), and it puts more burden on the client (it needs to reorder results and buffer them).

cc @HyukjinKwon @amaliujia @grundprinzip

Reply:
@cloud-fan and I had a discussion on this. I remember our initial thought was to maintain the partition ordering on the server side. For example, there is `def foreachPartition` API already. The question was whether we can foreach partitions in the partitioning order.

Reply:
Another question is do we always need to maintain the partition ordering?

Reply:
i guess the reordering in the client side is cheap, it only sort # batch  ints

Reply:
i guess the reordering in the client side is cheap, it only sort # batch  ints